ocaml/xapi/xapi_database_backup.ml

```diff
@@ -0,0 +1,357 @@
+module D=Debug.Make(struct let name="xapi_database_backup" end)
+open D
+
+open Http
+open Xmlrpc_client
+open Db_cache_types
+open Stdext
+open Threadext
+
+type stat =
+  { created :  int64
+  ; modified : int64
+  ; deleted :  int64
+  } [@@deriving rpc]
+
+type field =
+  { fldname : string
+  ; stat :    stat
+  ; value :   Schema.Value.t
+  } [@@deriving rpc]
+
+type row =
+  { objref : string
+  ; stat :   stat
+  ; fields :  field list
+  } [@@deriving rpc]
+
+type table =
+  { tblname : string
+  ; stat :    stat
+  ; rows :   row list
+  } [@@deriving rpc]
+
+type del_tables =
+  { key:   string
+  ; table: string
+  } [@@deriving rpc]
+
+type delta =
+  { fresh_token: int64
+  ; last_event_token: int64
+  ; tables :     table list
+  ; deletes :    del_tables list
+  ; counts : (string * int) list
+  } [@@deriving rpc]
+
+type update =
+  { stat:stat
+  ; tblname:string
+  ; objref:string
+  ; fldname:string
+  ; value:Schema.Value.t
+  }
+
+exception Content_length_required
+let delta_to_string (d:delta) = Jsonrpc.to_string (rpc_of_delta d)
+
+let get_gen db =
+  db
+  |> Database.manifest
+  |> Manifest.generation
+
+(* Master Functions *)
+
+let get_table_list db =
+  TableSet.fold_over_recent (-2L) (fun table stat value acc -> table::acc) (Database.tableset db) []
+
+let get_del_from_table_name token ts table =
+  Table.fold_over_deleted token (fun key _ acc -> ({key; table})::acc) (TableSet.find table ts) []
+
+let get_deleted ~token db =
+  List.concat (List.rev_map (get_del_from_table_name token (Database.tableset db)) (get_table_list db))
+
+let check_for_updates token db =
+  let ts = Database.tableset db in
+  let handle_obj obj = Row.fold_over_recent token (fun fldname {created; modified; deleted} value acc ->
+      {fldname; stat={created;modified;deleted}; value=value}::acc) obj [] in
+  let handle_table table = Table.fold_over_recent token (fun objref {created; modified; deleted} value acc ->
+      {objref; stat={created;modified;deleted}; fields=(handle_obj value)}::acc) table [] in
+  TableSet.fold_over_recent token (fun tblname {created; modified; deleted} value acc ->
+      {tblname; stat={created;modified;deleted}; rows=(handle_table value)}::acc) ts []
+
+let counter db table =
+  (table, (Table.fold (fun _ _ _ acc -> succ(acc)) (TableSet.find table (Database.tableset db)) 0))
+
+let object_count db =
+  List.rev_map (counter db) (get_table_list db)
+
+let get_delta __context token =
+  Mutex.execute Db_lock.dbcache_mutex (fun () ->
+      let db = Db_ref.get_database (Context.database_of __context) in
+      let tables = check_for_updates token db in
+      let deleted = get_deleted token db in
+      let t = Manifest.generation (Database.manifest db) in
+      {fresh_token=t;
+       last_event_token=Int64.pred t;
+       tables=tables;
+       deletes=deleted;
+       counts=object_count db})
```

Comment:
Both `object_count` and `get_deleted` internally call `get_table_list db`. We can pull that out and do it only once.